### PR TITLE
Fix thrust::optional<T&>::emplace()

### PR DESCRIPTION
**Pull request by Snektron**
## Description

closes https://github.com/NVIDIA/cccl/issues/1706

Where optional<T> inherits optional<T>::construct via a series of classes, optional<T&> does not. This means that optional<T&>::emplace() was broken and called into a member function that did not exist.

This replaces the functionality to make optional<T&>::emplace() change the stored reference to the new one. Note that it does _not_ emplace the referee, as this would lead to questionable behavior when the optional holds nullopt.

This was revealed by a change in LLVM, see
https://github.com/llvm/llvm-project/pull/90152 and https://github.com/ROCm/rocThrust/issues/404.

## Checklist

- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: thrust/thrust/optional.h [hunk 1]
```diff
@@ -2776,10 +2776,11 @@ class optional<T&>
...
   _CCCL_HOST_DEVICE T& emplace(Args&&... args) noexcept
   {
-    static_assert(std::is_constructible<T, Args&&...>::value, "T must be constructible with Args");
+    static_assert(sizeof...(Args) == 1, "Args must be a single lvalue");
+    static_assert(std::is_lvalue_reference<Args...>::value, "Args must be a single lvalue");
 
-    *this = nullopt;
-    this->construct(std::forward<Args>(args)...);
+    m_value = thrust::addressof(args...);
+    return **this;
```
Comment (miscco) [inline thrust/thrust/optional.h:2783]: I would simplify this to
```suggestion
    return *m_value;
```

File: thrust/testing/regression/gh_1706__thrust_optional_reference_emplace.cu [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#include <thrust/optional.h>
+#include <unittest/unittest.h>
+
+int main()
+{
+    int a = 10;
```
Comment (miscco) [inline thrust/testing/regression/gh_1706__thrust_optional_reference_emplace.cu:6]: Out of pure paranoia, please add a test that ensures we can emplace an empty optional

File: thrust/thrust/optional.h [hunk 3]
```diff
@@ -2776,10 +2776,11 @@ class optional<T&>
   template <class... Args>
   _CCCL_HOST_DEVICE T& emplace(Args&&... args) noexcept
   {
-    static_assert(std::is_constructible<T, Args&&...>::value, "T must be constructible with Args");
+    static_assert(sizeof...(Args) == 1, "Args must be a single lvalue");
+    static_assert(std::is_lvalue_reference<Args...>::value, "Args must be a single lvalue");
```
Comment (miscco) [inline thrust/thrust/optional.h:2780]: I would rather change the function signature to just take a `_Arg&` instead of variadics

**Review by miscco — COMMENTED** [review]
Thanks a lot for addressing the issue. I have some minor comments

**Comment by miscco** [comment]
pre-commit.ci autofix

File: thrust/thrust/optional.h (hunk 3 above)
Comment (Snektron) [reply thrust/thrust/optional.h]: That was my initial preference, but I didn't want to change the interface. I guess functionally the change would still be the same as changing the interface anyway, so it wouldn't matter technically.

**Review by miscco — APPROVED** [review]
Thanks a lot again :tada:
EJxuAEo

**Comment by Snektron** [comment]
@miscco Is the regressions directory the right place to put the test? I've noticed that they are disabled currently. I'm fine with moving, I wasn't sure what the proper place was otherwise.

**Comment by miscco** [comment]
I just realized, that there are no tests against thrust optional at all 🤦‍♂️
